agent/consul/leader_metrics.go

```diff
@@ -130,24 +134,35 @@ func (m CertExpirationMonitor) Monitor(ctx context.Context) error {
 
 	logger := m.Logger.With("metric", strings.Join(m.Key, "."))
 
+	fn := func() {
+		d, err := m.Query()
+		if err != nil {
+			logger.Warn("failed to emit certificate expiry metric", "error", err)
+			return
+		}
+
+		if d < 24*time.Hour {
+			logger.Warn("certificate will expire soon",
+				"time_to_expiry", d, "expiration", time.Now().Add(d))
+		}
+
+		expiry := d / time.Second
+		metrics.SetGaugeWithLabels(m.Key, float32(expiry), m.Labels)
+	}
+
+	// emit the metric immediately so that if a cert was just updated the
+	// new metric will be updated to the new expiration time.
+	fn()
+
 	for {
 		select {
```

Comment:
Also, I may be a bit naive here but `select`s without a `default` block make me a bit uneasy.

Theoretically, they could block forever, right?

> Since communication on nil channels can never proceed, a select with only nil channels and no default case blocks forever.

https://golang.org/ref/spec#Select_statements

Just trying to learn what the rationale might have been behind not including it originally.

Reply:
I believe it is pretty common to have no `default` in `select`, but it is definitely important to have a "shutcase" case. Here the shutdown case is `ctx.Done()`. The context will be cancelled by the caller when it wants to shutdown the goroutine. At that time the goroutine will exit. Using `ctx.Done()` should be a very common pattern in Go for managing the lifecycle of goroutines. It's unfortunately not yet super common in Consul, but we have been using it a lot more over the last year or so.

I believe the key point in those docs is "with only nil channels". Here both cases (`ctx.Done()`, and `ticker.C`) can not be nil, so I don't think we are at a risk of blocking forever.

Adding a `default` case actually changes the behaviour quite a bit. It makes it so that the select doesn't actually wait on the channels, but instead immediately goes to the `default` if there are no values ready in any of the channels. If we were to add a `default` here it would change this logic from a rather efficient select that blocks on operations, to a busy for loop that is constantly consuming CPU.

Reply:
thanks for the explanation 💯 !